ref: migrate streamEncryptedResponses flow to Typescript

karrui opened the pull request:
This PR is built upon #592.

## Problem
<!-- What problem are you trying to solve? What issue does this close? -->

This PR migrates the previous streamEncryptedResponses function in `encrypt-submissions.server.controller.js` to the correct layered architecture flow with Typescript.

## Solution
<!-- How did you solve the problem? -->

**Features**:

- add `EncryptedSubmissionModel#getSubmissionCursorByFormId` static method, migrating from the inlined code in the controller
- add `EncryptSubmissionService#getSubmissionCursor` service fn that abstracts away the static model method
- add `EncryptSubmissionService#transformAttachmentMetaStream` service fn to generate transform stream
- add `EncryptSubmissionController#handleStreamEncryptedResponses` handler function to replace old controller method

**Improvements**:

- add celebrate validation to check for valid start/end date query params


PS: No tests for the controller as I'm not sure how to test for pipes/streams and Google is not being very helpful. Will best performing integration tests when the router is migrated instead.

mantariksh commented on src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts:
```diff
@@ -0,0 +1,106 @@
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
```
use `StatusCodes`

mantariksh commented on src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts:
```diff
@@ -0,0 +1,106 @@
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
```
here too

mantariksh commented on src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts:
```diff
@@ -0,0 +1,106 @@
+        meta: logMeta,
+        error,
+      })
+      return res.status(500).json({
```
and here

mantariksh commented on src/app/routes/admin-forms.server.routes.js:
```diff
@@ -453,12 +454,24 @@ module.exports = function (app) {
+        .keys({
+          // Ensure YYYY-MM-DD format.
+          startDate: Joi.string().regex(
+            /([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/,
```
again, possible to extract this regex into a constant somewhere?

mantariksh commented on src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts:
```diff
@@ -0,0 +1,106 @@
+        meta: logMeta,
+      })
+
+      return res.end()
```
what's the difference between `end` and `close`? we previously didn't have `close` right? might there be errors if `res.end()` is called twice?

karrui replied:
That's a good question. From my refactoring and the additional tests, the `end` event was actually never emitted! So I added the close event handler. Maybe I'll remove the end handler since the close event should always be emitted at the end

liangyuanruo replied:
I don't think you should use [close](https://nodejs.org/api/stream.html#stream_event_end_1) since our code simply receives the stream object from mongoose, but isn't actually in control of the stream. I believe you'll also end up with 5XX errors if mongoose decides to close the stream and `res.end()` ends up being called twice.

karrui replied:
As discussed, end is not called, removed and use close instead.

Will create issue to allow for frontend to also stop waiting once the pipe is closed on the backend

liangyuanruo replied:
this should already be compatible with the frontend, so feel free to merge.

mantariksh reviewed (approved):
minor comments/questions

karrui commented:
Quite a number of merge conflicts due to https://github.com/opengovsg/FormSG/pull/555. Will request rereview once this is in 😆

karrui commented:
Ready for review. Also refactored @frankchn's s3 transform pipe process into typescript and added tests.

@frankchn would you mind reviewing it too?

frankchn reviewed (approved):
LGTM on the `transformAttachmentMetaStream` and related bits. Thanks for doing this!

mantariksh reviewed (approved):
fhrj
